clients/drcachesim/scheduler/flexible_queue.h

```diff
@@ -0,0 +1,219 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* A priority queue with constant-time search and removal from the middle. */
+
+#ifndef _DRMEMTRACE_FLEXIBLE_QUEUE_H_
+#define _DRMEMTRACE_FLEXIBLE_QUEUE_H_ 1
+
+/**
+ * @file drmemtrace/flexible_queue.h
+ * @brief DrMemtrace flexible priority queue.
+ */
+
+#define NOMINMAX // Avoid windows.h messing up std::max.
+#include <iostream>
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "utils.h"
+
+namespace dynamorio {  /**< General DynamoRIO namespace. */
+namespace drmemtrace { /**< DrMemtrace tracing + simulation infrastructure namespace. */
+
+/**
+ * A priority queue with constant-time search and removal from the middle.
+ */
+template <typename T, class comparator_t = std::less<T>> class flexible_queue_t {
+public:
+    typedef typename std::vector<T>::size_type index_t;
+    // Wrap max in parens to work around Visual Studio compiler issues with the
+    // max macro (even despite NOMINMAX defined above).
+    static constexpr index_t INVALID_INDEX = (std::numeric_limits<index_t>::max)();
+
+    flexible_queue_t() = default;
+    explicit flexible_queue_t(int verbose)
+        : verbose_(verbose)
+    {
+    }
+    bool
+    push(T entry)
+    {
+        if (entry2index_.find(entry) != entry2index_.end())
+            return false; // Duplicates not allowed.
+        entries_.push_back(entry);
+        index_t node = entries_.size() - 1;
+        entry2index_[entry] = node;
+        percolate_up(node);
+        print("after push");
+        return true;
+    }
+
+    void
+    pop()
+    {
+        erase(top());
+    }
+
+    T
+    top() const
+    {
+        return entries_[0]; // Undefined if empty.
+    }
+
+    bool
+    empty() const
+    {
+        return entries_.empty();
+    }
+
+    size_t
+    size() const
+    {
+        return entries_.size();
+    }
+
+    bool
+    find(T entry)
+    {
+        auto it = entry2index_.find(entry);
+        if (it == entry2index_.end())
+            return false;
+        return true;
+    }
+
+    bool
+    erase(T entry)
+    {
+        auto it = entry2index_.find(entry);
+        if (it == entry2index_.end())
+            return false;
+        index_t node = it->second;
+        if (node == entries_.size() - 1) {
+            entry2index_.erase(entry);
+            entries_.pop_back();
+            return true;
+        }
+        swap(node, entries_.size() - 1);
+        entry2index_.erase(entry);
+        entries_.pop_back();
+        percolate_down(node);
+        percolate_up(node);
+        print("after erase");
+        return true;
+    }
+
+private:
+    void
+    print(const std::string &message)
+    {
+        if (verbose_ <= 0)
+            return;
+        std::cout << message << "\n";
+        for (index_t i = 0; i < entries_.size(); ++i) {
+            std::cout << i << ": " << entries_[i] << " @ " << entry2index_[entries_[i]]
```

Comment:
This assumes T overloads <<. If it doesn't it'll cause a compile issue. Add to docx.

Reply:
I thought about throwing out all the printing code...but it may be useful in the future.  I added the << to the doc comment.